### PR TITLE
Introduce COUModel with waiting and auto-reconnection

### PR DESCRIPTION
**Pull request by rgildein**
- replace JujuWaiter with COUModel

### CODE REVIEW

File: cou/utils/juju_utils.py [hunk 1]
```diff
@@ -414,104 +424,68 @@ async def async_set_application_config(
...
+        """Make sure that model is connected."""
+        await self._model.disconnect()
+        await self._model.connect(model_name=self.name)
+
+    @retry(
+        wait=wait_random_exponential(multiplier=1, max=DEFAULT_TIMEOUT),
+        retry=retry_if_not_exception_type(BakeryException),
+        retry_error_callback=retry_error_callback("Model cloud not be connected."),
+    )
+    async def get_model(self) -> Model:
```
Comment (esunar) [inline cou/utils/juju_utils.py:472]: This should be private, no one should access model directly.

File: cou/utils/juju_utils.py [hunk 2]
```diff
@@ -414,104 +424,68 @@ async def async_set_application_config(
...
+    def disconnected(self) -> bool:
+        """Check if model is connected."""
+        return not (self._model.is_connected() and self._model.connection().is_open)
+
+    @property
+    def name(self) -> str:
+        """Return model name."""
+        return self._name
+
+    async def connect(self) -> None:
```
Comment (esunar) [inline cou/utils/juju_utils.py:462]: This should be private also.

File: cou/utils/juju_utils.py [hunk 3]
```diff
@@ -414,104 +424,68 @@ async def async_set_application_config(
...
+    """
+
+    def __init__(self, name: str):
+        """COU Model initialization with model_name compared to original juju.model.Model."""
+        self._model = Model(max_frame_size=JUJU_MAX_FRAME_SIZE)
+        self._name = name
+        self.logger = logging.getLogger(name)
+
+    @property
+    def disconnected(self) -> bool:
```
Comment (esunar) [inline cou/utils/juju_utils.py:453]: This should be private.

File: cou/utils/juju_utils.py [hunk 4]
```diff
@@ -414,104 +424,68 @@ async def async_set_application_config(
...
+        await self._model.connect(model_name=self.name)
+
+    @retry(
+        wait=wait_random_exponential(multiplier=1, max=DEFAULT_TIMEOUT),
+        retry=retry_if_not_exception_type(BakeryException),
+        retry_error_callback=retry_error_callback("Model cloud not be connected."),
+    )
+    async def get_model(self) -> Model:
+        """Get juju.model.Model and make sure that's it's connected."""
+        if not self.disconnected:
```
Comment (esunar) [inline cou/utils/juju_utils.py:474]:
```suggestion
        if self.disconnected:
```

**Comment by rgildein** [comment]
superseded by #53

**Pull request closed without merging**